iceoryx2-ffi/cxx/include/iox2/header_publish_subscribe.hpp

```diff
@@ -30,8 +30,8 @@ class HeaderPublishSubscribe {
     /// Returns the [`UniquePublisherId`] of the source [`Publisher`].
     auto publisher_id() const -> UniquePublisherId;
 
-    /// Returns the [`Layout`] of the corresponding payload.
-    auto payload_type_layout() const -> iox::Layout;
+    /// Returns the number of [`Payload`] elements in the received [`Sample`].
+    auto number_of_elements() const -> uint64_t;
```

Comment:
I think the layout is additionally needed here.

With these changes, the code to send slice copies in `iceoryx-ffi` looks like:

```
unsafe fn send_slice_copy<S: Service>(
    publisher: &Publisher<S, PayloadFfi, UserHeaderFfi>,
    data_ptr: *const c_void,
    size_of_element: usize,
    number_of_elements: usize,
    number_of_recipients: *mut usize,
) -> c_int {
    // loan_slice_uninit(1) <= 1 is correct here since it defines the number of
    // slice elements not bytes. The element was set via TypeDetails and has a
    // defined size and alignment.
    let mut sample = match publisher.loan_custom_payload(number_of_elements) {
        Ok(sample) => sample,
        Err(e) => return e.into_c_int(),
    };

    if sample.payload().len() < size_of_element * number_of_elements {
        return iox2_publisher_send_error_e::LOAN_ERROR_EXCEEDS_MAX_LOAN_SIZE as c_int;
    }

    let sample_ptr = sample.payload_mut().as_mut_ptr();
    core::ptr::copy_nonoverlapping(data_ptr, sample_ptr.cast(), size_of_element);
    match sample.assume_init().send() {
        Ok(v) => {
            if !number_of_recipients.is_null() {
                *number_of_recipients = v;
            }
        }
        Err(e) => return e.into_c_int(),
    }

    IOX2_OK
}
```

Focusing on this part:

```
    if sample.payload().len() < size_of_element * number_of_elements {
        return iox2_publisher_send_error_e::LOAN_ERROR_EXCEEDS_MAX_LOAN_SIZE as c_int;
    }
```

This code is errorneous because it does not consider the alignment of the element. There could be an alignment set which requires more bytes than loaned by the publisher.

Alternatively, I guess the alignment could be passed in from as an argument...

Reply:
I went with the alternative option, see: https://github.com/eclipse-iceoryx/iceoryx2/pull/492/commits/5411dd69d5c42a34f7585da7f0388b50347ea1d2#diff-1e40fb692713e95d0a2622c1ae500770e1954f70ddec5d2ce1e7bf0a48d40327R227

If this is OK, then nothing to change here.